torch/_prims/nvfuser_executor.py

```diff
@@ -161,6 +166,15 @@ def call_function(self, target, args, kwargs):
                 args = (fd,) + args
                 return target(*args, **kwargs)
 
+            def output(self, target, args, kwargs):
```

Comment:
An alternative to `def output` is to put `if node.op == "output":` in `run_node` method. Then the node information is available and we wouldn't need to save the output node before defining the Interpreter. Not that it matters much, so no action is needed.

Reply:
>  we wouldn't need to save the output node before defining the Interpreter.

It's still easier that we save outputs before hand. When we reach `output` node in the interpreter, we have already transformed all tensors to `TensorView` and lost the original tensor with meta information.

Reply:
Here's an example of how meta information saved in the GraphModule is used in the middle of the interpreter call to insert the "original_shape" argument needed for `squeeze`:
https://github.com/csarofeen/pytorch/blob/cfdabc1b5a18b7fb11619af295ac53b8e21afda0/torch/_prims/nvfuser_executor.py#L130-L138

Reply:
IIUC, you are pointing me to this line `original_shape = list(node.args[0].meta["tensor_meta"].shape)`?

We'll still need to do this outside of the interpreter though. Because this interpreter runs on `nv_args` instead of pytorch tensor vvv.
```
        nv_args = tuple(map(templates_to_nvfuser_inputs, nv_args_templates))
        out = FusionInterpreter(gm).run(*nv_args)
```

Which means `node.args[x]` would give us dtype of the new codegen tensor, and that has no reference value any more. :cry: We need to look at the original dtype from type propagation done with fake tensor (Not totally sure who did it actually).

Reply:
Yes, the interpreter runs on `nv_args`, but no new nodes are created, it's all the original recorded graph.
`node.args` inside `run_node` would give exactly the same as was accessed with:
```py
for node in gm.graph.node:
    print(node.args)
```
`args, kwargs = gm.fetch_args_kwargs_from_env(node)` would give the new thing that has no reference to the original graph though.

Here's a concrete example:
```py
import torch
from torch.fx.experimental.proxy_tensor import make_fx

class TestInterpreter(torch.fx.Interpreter):
    def run_node(self, node):
        if node.op == "output":
            saved_args_dtypes = list(arg.meta["tensor_meta"].dtype for arg in node.args)
            print(f"TestInterpreter: args_dtypes: {saved_args_dtypes}")
            args, kwargs = self.fetch_args_kwargs_from_env(node)
            fetched_args_dtypes = list(arg.dtype for arg in args)
            print(f"TestInterpreter: fetched_args_dtypes: {fetched_args_dtypes}")
            return self.output(node.target, args, kwargs)
        return super().run_node(node)

a = torch.randn(3, 3, dtype=torch.float16)

def func(a):
    return a + 1

gm = make_fx(func)(a)
gm.graph.print_tabular()

output_node = next(filter(lambda n: n.op == "output", gm.graph.nodes))
print(output_node.args[0].meta["tensor_meta"]) # {'dtype': torch.float16}

TestInterpreter(gm).run(a.double())
TestInterpreter(gm).run(a.float())
TestInterpreter(gm).run(a.cdouble())
TestInterpreter(gm).run(a.cfloat())
# Output:
# TestInterpreter: saved_args_dtypes: [torch.float16]
# TestInterpreter: fetched_args_dtypes: [torch.float64]
# TestInterpreter: saved_args_dtypes: [torch.float16]
# TestInterpreter: fetched_args_dtypes: [torch.float32]
# TestInterpreter: saved_args_dtypes: [torch.float16]
# TestInterpreter: fetched_args_dtypes: [torch.complex128]
# TestInterpreter: saved_args_dtypes: [torch.float16]
# TestInterpreter: fetched_args_dtypes: [torch.complex64]
```

Reply:
I see how this works now... I was previously doing things inside `def output`, which passes arguments from retrieved with `fetch_args_kwargs_from_env`. So I was under the impression that the graph is mutated during the run. Should have paid more attention to your previous comment. :face_with_head_bandage: 

Thanks a lot for bearing with my ignorance. It is a cleaner alternative. Let me refactor the code in upstream.